tslearn/metrics.py

```diff
@@ -653,21 +654,28 @@ def cdist_dtw(dataset1, dataset2=None, global_constraint=None,
     """
     dataset1 = to_time_series_dataset(dataset1)
 
-    if global_constraint is not None:
-        global_constraint_str = global_constraint
-    else:
-        global_constraint_str = ""
-
     if dataset2 is None:
+        if global_constraint is not None:
+            global_constraint_str = global_constraint
+        else:
+            global_constraint_str = ""
+
+        # TODO
         return njit_cdist_dtw_self(
```

Comment:
Can't we do something similar here (I'm assuming we want pairwise distances between all timeseries within `dataset1`). Perhaps here we could just say `dataset2 = dataset1` and then always call the code of L670-677

Reply:
When computing pairwise distances between all timeseries within `dataset1`, the matrix is symmetric and all the elements on the diagonale are zeros. There are only `n * (n-1) / 2` values to compute instead of  `n * n`.

Reply:
The idea here is to save half the computations using the fact that dtw(x,y)=dtw(y,x) and dtw(x,x)=0

Reply:
Hmm ok, it would still be possible to call the Parallel code on this right?

```
matrix = np.zeros((len(dataset1), len(dataset2)))
matrix[np.triu_indices(len(dataset(1), m=len(dataset2))] = Parallel(n_jobs=n_jobs)(
            delayed(dtw)(
                dataset1[i], dataset2[j],
                global_constraint=global_constraint,
                sakoe_chiba_radius=sakoe_chiba_radius,
                itakura_max_slope=itakura_max_slope)
            for i in range(len(dataset1)) for j in range(i + 1, len(dataset2))
        )
matrix = matrix + matrix.T
```

Reply:
This is exactly the kind of code I was looking for, thanks ! I will give it a try.

Reply:
Please open a new issue on this specific point describing the context and the error you get, so that we can see what could be done.